pkg/controller/volume/scheduling/scheduler_binder.go

```diff
@@ -836,6 +881,75 @@ func (b *volumeBinder) revertAssumedPVCs(claims []*v1.PersistentVolumeClaim) {
 	}
 }
 
+// hasEnoughCapacity checks whether the provisioner has enough capacity left for a new volume of the given size
+// that is available from the node.
+func (b *volumeBinder) hasEnoughCapacity(provisioner string, resources *v1.ResourceRequirements, storageClass *storagev1.StorageClass, node *v1.Node) (bool, error) {
+	// This is an optional feature. If disabled, we assume that
+	// there is enough storage.
+	if !b.capacityCheckEnabled {
+		return true, nil
+	}
+
+	quantity, ok := resources.Requests[v1.ResourceStorage]
+	if !ok {
+		// No capacity to check for.
+		return true, nil
+	}
+
+	// Only enabled for CSI drivers which opt into it.
+	driver, err := b.csiDriverLister.Get(provisioner)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Either the provisioner is not a CSI driver or the driver does not
+			// opt into storage capacity scheduling. Either way, skip
+			// capacity checking.
+			return true, nil
+		}
+		return false, err
+	}
+	if driver.Spec.StorageCapacity == nil || !*driver.Spec.StorageCapacity {
+		return true, nil
+	}
+
+	// Look for a matching CSIStorageCapacity object(s).
+	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure.
```

Comment:
some thoughts: indexing by storage class should be valuable as the class field is immutable
we also can index capacity objects by nodes as the labels normally do not change frequently

Reply:
lets create an issue to track all optimization work and use it in the todo here.

Reply:
I've started a TODO list in the enhancement issue: https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718

It's also part of the beta graduation criteria in the KEP, but it makes sense to track specific TODOs like this one so that they don't get forgotten.